docs/readme-zephyr.md

```diff
@@ -156,3 +162,50 @@ although MCUboot's key management infrastructure supports multiple keypairs.
 
 Once MCUboot is built, this new keypair file (`mykey.pem` in this
 example) can be used to sign images.
+
+## Using swap-using-scratch flash algorithm
+
+To use the swap-using-scratch flash algorithm, a scratch partition needs to be
+present for the target board which is used for holding the data being swapped
+from both slots, this section must be at least as big as the largest sector
+size of the 2 partitions (e.g. if a device has a primary slot in main flash
+with a sector size of 512 bytes and secondar slot in external off-chip flash
+with a sector size of 4KB then the scratch area must be at least 4KB in size).
+The number of sectors must also be evenly divisable by this sector size, e.g.
+4KB, 8KB, 12KB, 16KB are allowed, 7KB, 7.5KB are not. This scratch partition
+needs adding to the .dts file for the board, e.g. for the nrf52dk_nrf52832
+board thus would involve updating
+`<zephyr>/boards/arm/nrf52dk_nrf52832/nrf52dk_nrf52832.dts` with:
+
+```
+    boot_partition: partition@0 {
+        label = "mcuboot";
+        reg = <0x00000000 0xc000>;
+    };
+    slot0_partition: partition@c000 {
+        label = "image-0";
+        reg = <0x0000C000 0x37000>;
+    };
+    slot1_partition: partition@43000 {
+        label = "image-1";
+        reg = <0x00043000 0x37000>;
+    };
+    storage_partition: partition@7a000 {
+        label = "storage";
+        reg = <0x0007a000 0x00006000>;
+    };
+```
+
+Which would make the application size 220KB and scratch size 24KB (the nRF52832
+has a 4KB sector size so the size of the scratch partition can be reduced at
+the cost of vastly reducing flash lifespan, e.g. for a 32KB firmware update
+with an 8KB scratch area, the scratch area would be erased and programmed 8
+times per image upgrade/revert). To configure MCUboot to work in
+swap-using-scratch mode, the Kconfig value must be set when building it:
+`CONFIG_BOOT_SWAP_USING_SCRATCH=y`.
+
+Note that it is possible for an application to get into a stuck state when
+swap-using-scratch is used whereby an application has loaded a firmware update
+and marked it as test/confirmed but MCUboot will not swap the images and
+erasing the secondary slot from the zephyr application returns an error
+because the slot is marked for upgrade.
```

Comment:
Is this related to #1165? Was anyone else able to reproduce it or you were the only person that has ever seen the error? Does it also happen when using only internal flash?

Reply:
Everyone at the company I worked at on multiple boards (pinnacle 100, bt610), people on discord with varying boards. Both boards I've seen it on had primary slots in internal flash, with secondary and scratch partitions in external QSPI memory (all devices with 4KB sector sizes)

Reply:
Pinnacle 100 and bt610 are also nrf52840, correct? Do they also use QSPI flash? MCUboot doesn't really know if the flash is internal or external, that's up to the OS backend, but I am wondering if there is some pattern here. How many upgrades in average I would have to run to see the issue? Do you know if uploading new firmware with MCUmgr in between tests is necessary to trigger the issue, or just doing test/revert of the same images is enough?

Reply:
Not sure on that, it was completely random as to when it happened, as bad as it sounds there was no easily reproducible method to get a device into this state (it was all during development, might have been reported post production too but I do not remember) - images would have been loaded via MCUmgr (bluetooth and/or shell transports) then marked as test from that, the image wasn't left and flipped between test and none.

Reply:
Do we need to create an issue to track this? Ideally, we would be able to figure out how to reproduce this in the simulator. Does it happen only with the write sizes differing between internal and external? But, I don't think creating the issue should block merging this issue. Let me know in a comment that there is an issue created, and this can proceed.

Reply:
I gave it a quick try to reproduce it and was not able to in 20 minutes or so. The write and erase sizes for both memories were the same (mx25... QSPI flash - 4KB sector/erase size, 4 byte minimum write size, nrf52840 flash - 4KB sector/erase size, 4 byte minimum write size), and it was completely random as to how it got into that mode, but once it was in that mode, you were unable to get it out of that mode or load a new update, mcuboot would not do any swapping and the app would not allow uploading a new image because the image was marked for test

Reply:
Yes, please open an issue with the same description. I have used swap with scratch and newtmgr together for many years with Mynewt, and to some extent with Zephyr as well, and never bumped into this bug. I am somewhat suspicious it might be related to the QSPI driver, like an error write that does not raise an error or something similar, but it's just a guess. Btw, I've never tried scratch on the QSPI flash, but for MCUboot flash is just flash, wherever it is located...